bpfd/src/command.rs

```diff
@@ -484,77 +382,84 @@ impl Program {
-                p.info.if_index,
-                Some(p.info.direction),
+                // if_index should be resolved by this point
+                p.if_index.unwrap(),
+                Some(p.direction),
             ))),
             _ => None,
         }
     }
 
+    // TODO(astoycos) move this to `ProgramData` specific methods
     pub(crate) fn data_mut(&mut self) -> &mut ProgramData {
         match self {
-            Program::Xdp(p) => &mut p.data,
-            Program::Tracepoint(p) => &mut p.data,
-            Program::Tc(p) => &mut p.data,
-            Program::Kprobe(p) => &mut p.data,
-            Program::Uprobe(p) => &mut p.data,
+            Program::Xdp(p) => Some(&mut p.data),
+            Program::Tracepoint(p) => Some(&mut p.data),
+            Program::Tc(p) => Some(&mut p.data),
+            Program::Kprobe(p) => Some(&mut p.data),
+            Program::Uprobe(p) => Some(&mut p.data),
+            Program::Unsupported(_) => None,
         }
+        .expect("Unsupported maps have no bpfd specific data")
     }
 
+    // TODO convert to using data_op everwhere, the extra unwrap is just a bit
+    // cumbersome.
     pub(crate) fn data(&self) -> &ProgramData {
         match self {
-            Program::Xdp(p) => &p.data,
-            Program::Tracepoint(p) => &p.data,
-            Program::Tc(p) => &p.data,
-            Program::Kprobe(p) => &p.data,
-            Program::Uprobe(p) => &p.data,
+            Program::Xdp(p) => Some(&p.data),
+            Program::Tracepoint(p) => Some(&p.data),
+            Program::Tc(p) => Some(&p.data),
+            Program::Kprobe(p) => Some(&p.data),
+            Program::Uprobe(p) => Some(&p.data),
+            Program::Unsupported(_) => None,
         }
+        .expect("Unsupported maps have no bpfd specific data")
     }
 
-    pub(crate) fn metadata(&self) -> Option<&Metadata> {
+    pub(crate) fn data_op(&self) -> Option<&ProgramData> {
         match self {
-            Program::Xdp(p) => Some(&p.info.metadata),
-            Program::Tracepoint(_) => None,
-            Program::Tc(p) => Some(&p.info.metadata),
-            Program::Kprobe(_) => None,
-            Program::Uprobe(_) => None,
+            Program::Xdp(p) => Some(&p.data),
+            Program::Tracepoint(p) => Some(&p.data),
+            Program::Tc(p) => Some(&p.data),
+            Program::Kprobe(p) => Some(&p.data),
+            Program::Uprobe(p) => Some(&p.data),
+            Program::Unsupported(_) => None,
         }
     }
 
-    pub(crate) fn owner(&self) -> &String {
+    pub(crate) fn attached(&self) -> bool {
         match self {
-            Program::Xdp(p) => &p.data.owner,
-            Program::Tracepoint(p) => &p.data.owner,
-            Program::Tc(p) => &p.data.owner,
-            Program::Kprobe(p) => &p.data.owner,
-            Program::Uprobe(p) => &p.data.owner,
+            Program::Xdp(p) => p.attached,
+            Program::Tc(p) => p.attached,
+            _ => false,
```

Comment:
I mean technically all TC, XDP, TRACEPOINT, UPROBE, and KPROBE can be "attached" right? For now I'll make this an option but I think we might possibly want to rethink this in the future.

Reply:
The point here is that, unless you understand the finer details of our internal API it's an easy footgun for a contributor to think that `Program::attached()` returns whether a program is attached or not.
It's not clear that this field is only used for tracepoint/xdp programs which to would indicate that we pick either:
a. It should be on the speciailzed program type (XDP/Tracepoint)
b. It's an Option, because that makes you have to handle the None case.